src/ipss_cell.hpp

```diff
@@ -0,0 +1,201 @@
+#include <util/unique_any.hpp>
+
+
+namespace arb {
+
+/// Cell implementation of inhomogeneous Poisson spike generator.
+/// Spikes are generated at a configurable sample rate. With a rate that can be
+/// varied across time. Rate can be piece wise constant or linearly interpolated
+/// per sample time step.
+/// All times supplied to the cell will be 'rounded' to the first higher multiple of
+/// of the supplied sample_delta.
+/// When interpolating the rate will be the averaged rate between the
+/// t and t+1. The return spike time will be t.
+/// The first rate supplied in the time-rate vector should be at or before the
+/// start time of the cell. The rate of the last supplied time-rate will be kept
+/// until the stop_time of the cell
+
+class ipss_cell   {
+public:
+    /// Constructor
+    ipss_cell(ipss_cell_description desc, cell_gid_type gid) :
+        gid_(gid), sample_delta_(desc.sample_delta),
+        start_step_(time_to_step_idx(desc.start_time, sample_delta_)),
+        stop_step_(time_to_step_idx(desc.stop_time, sample_delta_)),
+        current_step_(0), interpolate_(desc.interpolate),
+        generator_(gid_)
+    {
+        // For internal storage convert the rates from spikes / second to
+        // spikes per sample_delta_  and the time to step_idx from t = 0.0
+        for (auto entry : desc.rates_per_time) {
+            auto spikes_per_sample_delta = (entry.second / 1000.0) * sample_delta_;
+            auto time_step = time_to_step_idx(entry.first, sample_delta_);
+            rates_per_step_.push_back({ time_step, spikes_per_sample_delta });
+        }
+
+        // Add and extra rate entry at max long, copy of the last original entry
+        // This allows current and next iterator.
+        rates_per_step_.push_back({ -1, rates_per_step_.cend()->second });
+
+        // Sanity checks
+        if (start_step_ > stop_step_) {
+            throw std::logic_error("The start time is after the end time in the Inhomogeneous Poisson Spike Source.");
+        }
+
+        if (start_step_ < rates_per_step_.cbegin()->first) {
+            throw std::logic_error("The start time of the Inhomogeneous Poisson Spike Source is before the first time/rate pair");
+        }
+
+        reset();
+    }
+
+    cell_kind get_cell_kind() const {
+        return cell_kind::inhomogeneous_poisson_spike_source;
+    }
+
+    /// reset cell to start state
+    ///   - Find the first valid rate depending on the start time_
+    ///   - Set the rates based on the rate change iterators
+    ///   - Reseed the rng
+    void reset() {
+
+        // set the pointer to the first and second entry in the rates vector
+        current_rate_it_ = rates_per_step_.cbegin();
+        next_rates_it_ = current_rate_it_;
+        next_rates_it_++;
+        update_rates();
+
+        // Step through the rate_change vector if begin_step does not fall in the
+        // range of the current pointers
+        while (start_step_ >= next_rates_it_->first) {
```

Comment:
Can't we fall of the end here by incrementing `next_rates_it` or are we guaranteed to find `start_step_`?